elephant_vending_machine/libraries/vending_machine.py

```diff
@@ -0,0 +1,192 @@
+        if quit_selection:
+            self.pool.terminate()
+
+    def wait_for_input(self, groups, timeout):
+        """Waits for input on the motion sensors. If no motion is detected by the specified
+        time to wait, returns with a result to indicate this.
+
+        Returns:
+            selection: A string with value 'left', 'middle', 'right', or 'timeout', indicating
+            the selection or lack thereof.
+        """
+        self.result = None
+        start_time = time.perf_counter()
+        selection = 'timeout'
+        self.pool = mp.Pool()
+        for group in groups:
+            self.pool.apply_async(func=worker, args=(
+                group.get_group_id, self.addresses), callback=self.callback)
+        while self.result is None:
+            current_time = time.perf_counter()
+            if current_time - start_time > timeout:
+                self.pool.terminate()
+                break
+        self.pool.close()
+        self.pool.join()
+        if self.result == LEFT_SCREEN:
+            selection = 'left'
+        elif self.result == MIDDLE_SCREEN:
+            selection = 'middle'
+        elif self.result == RIGHT_SCREEN:
+            selection = 'right'
+        return selection
+
+
+class SensorGrouping:
+    """Provides an abstraction of the devices controlled by Raspberry Pis.
+
+    Pi's will have an LED strip, a distance measurement device, and a screen.
+    This class will provide utilities for interacting with individual sensors.
+    """
+
+    # This class is only one attribute over the recommended limit and all are needed.
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, address, screen_identifier, config=None):
+        """Initialize an instance of VendingMachine.
+
+        Provides an interface to the different sensors in divided by their associated screen.
+
+         Parameters:
+            address (str): The local network address of the Raspberry Pi controlling the sensors
+            screen_identifier (int): A integer
+            config (dict): The configuration values of the Flask server
+        """
+        self.factory = PiGPIOFactory(host=address)
+        self.group_id = screen_identifier
+        self.sensor = MotionSensor(MOTION_PIN, pin_factory=self.factory)
+        self.led = LED(LED_PIN, pin_factory=self.factory)
+        self.correct_stimulus = False
+        self.address = address
+        self.config = config
+        self.pid_of_previous_display_command = None
+
+    def change_led(self, color):
+        """Controls the lighting of LEDs for a given sensor grouping.
+
+        Parameters:
+            state (int): The desired state of the LED.
+        """
+        shell = spur.SshShell(
+            hostname=self.address,
+            username='pi',
+            missing_host_key=spur.ssh.MissingHostKey.accept,
+            load_system_host_keys=False
+        )
+        with shell:
+            shell.run(
```

Comment:
run() waits for the remote command to finish executing, so this will wait until the python process exits and then continue ([documentation can be found here](https://pypi.org/project/spur/)). spawn() starts a process and doesn't wait for it to exit. As far as image display, I'm using spawn() and saving the pid so that it can be killed, however I'm not doing this yet. Do you think I should add a close() method to the vending machine that will kill all image display processes as a neat way of cleaning them up? Or kill the previous process when we go to display a new image?

Reply:
I see what you mean with the treats. For the LEDs I could use spawn() as I'm doing with image display so that we don't have to wait for the LED script to return, and we'd then be able to immediately vend treats (I'll stub in a method for that and note it in the docs as a todo for the next group)

Reply:
This was changed so that the led_color_with_time method accepts an int representing the number of seconds the LEDs should be illuminated for. The LED script is called with shell.spawn, which means the experiment will not block until the LED script is finished running.